### PR TITLE
Format `PatternMatchStar`

### LINKED ISSUE #6558: Format `PatternMatchStar`
This would be done in a similar manner as mentioned here (https://github.com/astral-sh/ruff/issues/6555#issuecomment-1677241864) except that this would be delegated to [`Identifier`](https://github.com/astral-sh/ruff/blob/b05574babd6224af99938a429d0b19ba3297513c/crates/ruff_python_formatter/src/other/identifier.rs) formatting instead.

Implement formatting for the `*pat` pattern

```python
match x:
    case [1, 2, *rest]:
        ...
    case [*_]:
        ...
```

### PR DESCRIPTION
**Pull request by harupy**
## Summary

Close #6558

### CODE REVIEW

**Comment by harupy** [comment]
As a test, I formatted this code:

```python
x = [1, 2, 3, 4]

match x:
    case [
        1,
        *
        rest,
    ]:
        print(rest)
    case [
        1,
        2,
        *
        # hello
        rest,
    ]:
        print(rest)
    case [*_]:
        print("no match")
```

and got:

```python
x = [1, 2, 3, 4]

match x:
    case [NOT_YET_IMPLEMENTED_PatternMatchSequence, 2]:
        print(rest)
    case [NOT_YET_IMPLEMENTED_PatternMatchSequence, 2]:
        print(rest)
    case [NOT_YET_IMPLEMENTED_PatternMatchSequence, 2]:
        print("no match")
```

Does this mean #6645 needs to be resolved first?

**Comment by harupy** [comment]
Tweaked the PatternMatchSequence impl. and got this:

```python
x = [1, 2, 3, 4]

match x:
    case [
        "NOT_YET_IMPLEMENTED_PatternMatchValue",
        *rest,
    ]:
        print("a")
        print(rest)
    case [
        "NOT_YET_IMPLEMENTED_PatternMatchValue",
        "NOT_YET_IMPLEMENTED_PatternMatchValue",
        *rest# hello
        ,
    ]:
        print(rest)
    case [*_]:
        print("no match")

```

File: crates/ruff_python_formatter/src/pattern/pattern_match_star.rs [hunk 1]
```diff
@@ -1,19 +1,22 @@
...
     fn fmt_fields(&self, item: &PatternMatchStar, f: &mut PyFormatter) -> FormatResult<()> {
-        write!(
-            f,
-            [not_yet_implemented_custom_text(
-                "*NOT_YET_IMPLEMENTED_PatternMatchStar",
-                item
-            )]
-        )
+        let comments = f.context().comments().clone();
+        let dangling = comments.dangling_comments(item);
```
Comment (harupy) [inline crates/ruff_python_formatter/src/pattern/pattern_match_star.rs:14]: Should comments be handled in `PatternMatchSequence`?

**Comment by MichaReiser** [comment]
Hmm yeah, it may be challenging to test this just now without match as, singleton formatting, or sequence formatting. There are PRs for `as` and `singleton` but it may be good to do sequence first.

File: crates/ruff_python_formatter/src/pattern/pattern_match_star.rs (hunk 1 above)
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/pattern/pattern_match_star.rs:14]: It depends on the comments. What's the code snipped that requires the dangling comment handling?

Comment (harupy) [reply crates/ruff_python_formatter/src/pattern/pattern_match_star.rs:14]: @MichaReiser 

For example:

```python
x = [1, 2, 3, 4]

match x:
    case [
        1,
        *
        # comment here
        rest,
    ]:
        print(rest)
```

**Comment by harupy** [comment]
> Hmm yeah, it may be challenging to test this just now without match as, singleton formatting, or sequence formatting. There are PRs for `as` and `singleton` but it may be good to do sequence first.

Can I tackle #6645 first?

**Comment by MichaReiser** [comment]
> > Hmm yeah, it may be challenging to test this just now without match as, singleton formatting, or sequence formatting. There are PRs for `as` and `singleton` but it may be good to do sequence first.
> 
> Can I tackle #6645 first?

Sure. You'll have to comment on the issue so that I can assign it to you.

File: crates/ruff_python_formatter/resources/test/fixtures/ruff/statement/match.py [hunk 2]
```diff
@@ -236,3 +236,15 @@ def foo():
...
+    case [1, 2, *rest]:
+        pass
+    case [1, 2, *_]:
+        pass
+    case [
+        1,
+        2,
+        *rest,
+    ]:
+        pass
```
Comment (konstin) [inline crates/ruff_python_formatter/resources/test/fixtures/ruff/statement/match.py:250]: Could you add test case for the following?
```python
    case [1, 2, * # comment
        rest]:
        pass
    case [1, 2, * # comment
        _]:
        pass
```

File: crates/ruff_python_formatter/resources/test/fixtures/ruff/statement/match.py [hunk 3]
```diff
@@ -38,12 +38,12 @@
...
-    yield foo
-):  # another comment
-    case ["a", "b", "c"]:
-        pass
+# match (  # comment
+#     # let's go
+#     yield foo
+# ):  # another comment
+#     case ["a", "b", "c"]:
+#         pass
```
Comment (konstin) [inline crates/ruff_python_formatter/resources/test/fixtures/ruff/statement/match.py:46]: This should be uncommented before merging

**Review by konstin — APPROVED** [review]

File: crates/ruff_python_formatter/resources/test/fixtures/ruff/statement/match.py (hunk 3 above)
Comment (harupy) [reply crates/ruff_python_formatter/resources/test/fixtures/ruff/statement/match.py]: Forgot to uncomment this, thanks for the catch!

File: crates/ruff_python_formatter/resources/test/fixtures/ruff/statement/match.py (hunk 2 above)
Comment (harupy) [reply crates/ruff_python_formatter/resources/test/fixtures/ruff/statement/match.py:250]: Added!

File: crates/ruff_python_formatter/resources/test/fixtures/ruff/statement/match.py [hunk 4]
```diff
@@ -236,3 +236,21 @@ def foo():
...
+        2,
+        *rest,
+    ]:
+        pass
+    case [1, 2, * # comment
+        rest]:
+        pass
+    case [1, 2, * # comment
+        _]:
+        pass
```
Comment (harupy) [inline crates/ruff_python_formatter/resources/test/fixtures/ruff/statement/match.py:256]: This is formatted to:

```
    case [
        "NOT_YET_IMPLEMENTED_PatternMatchValue",
        "NOT_YET_IMPLEMENTED_PatternMatchValue",
        *rest  # comment
        ,
    ]:
        pass
    case [
        "NOT_YET_IMPLEMENTED_PatternMatchValue",
        "NOT_YET_IMPLEMENTED_PatternMatchValue",
        *_  # comment
        ,
    ]:
        pass
```

then formatted to:

```
    case [
        "NOT_YET_IMPLEMENTED_PatternMatchValue",
        "NOT_YET_IMPLEMENTED_PatternMatchValue",
        *rest,  # comment
    ]:
        pass
    case [
        "NOT_YET_IMPLEMENTED_PatternMatchValue",
        "NOT_YET_IMPLEMENTED_PatternMatchValue",
        *_,  # comment
    ]:
        pass
```

Comment (konstin) [reply crates/ruff_python_formatter/resources/test/fixtures/ruff/statement/match.py:256]: If you look at
```python
match foo:
    case [* # comment
        rest, 1]:
        pass
    case [*rest # comment
          , 1]:
        pass
```
you get
```
{
    Node {
        kind: PatternMatchStar,
        range: 21..45,
        source: `* # comment⏎`,
    }: {
        "leading": [],
        "dangling": [
            SourceComment {
                text: "# comment",
                position: EndOfLine,
                formatted: true,
            },
        ],
        "trailing": [],
    },
    Node {
        kind: PatternMatchStar,
        range: 74..79,
        source: `*rest`,
    }: {
        "leading": [],
        "dangling": [],
        "trailing": [
            SourceComment {
                text: "# comment",
                position: EndOfLine,
                formatted: true,
            },
        ],
    },
}
```
so i think the best solution is to manually reassign the comment from dangling to trailing.

Comment (charliermarsh) [reply crates/ruff_python_formatter/resources/test/fixtures/ruff/statement/match.py:260]: @harupy - I typically run something like `cargo run foo.py --emit=stdout --print-comments` from `ruff/crates/ruff_python_formatter`.

Comment (charliermarsh) [reply crates/ruff_python_formatter/resources/test/fixtures/ruff/statement/match.py:260]: Oh sorry. You're asking how to modify these assignments. Any such modifications need to be done in `placement.rs` -- those are essentially custom rules that let us change the association of comments. See, e.g., `handle_trailing_expression_starred_star_end_of_line_comment` which is somewhat similar.

Comment (konstin) [reply crates/ruff_python_formatter/resources/test/fixtures/ruff/statement/match.py:260]: You have to add a case in `handle_enclosed_comment` for when the enclosing node is `PatternMatchStar`. The general logic to fix comment placement is in `placement.rs`/`place_comment`.

File: crates/ruff_python_formatter/src/comments/placement.rs [hunk 5]
```diff
@@ -1187,6 +1188,24 @@ fn handle_pattern_match_as_comment<'a>(
...
+/// case [
+///   ...,
+///   *  # dangling end of line comment
+///   # dangling end of line comment
+///   rest,
+/// ]: ...
+/// ```
+fn handle_pattern_match_star_comment(comment: DecoratedComment) -> CommentPlacement {
+    debug_assert!(comment.enclosing_node().is_pattern_match_star());
+    let AnyNodeRef::PatternMatchStar(range, ..) = comment.enclosing_node() else {
```
Comment (konstin) [inline crates/ruff_python_formatter/src/comments/placement.rs:1203]: You can get the `ast::PatternMatchStar` from the `AnyNodeRef::PatternMatchStar(_)` in the match and then pass it to the function (like `handle_leading_class_with_decorators_comment`)

Comment (harupy) [reply crates/ruff_python_formatter/src/comments/placement.rs:1203]: Ah I see, this doesn't work.

Comment (charliermarsh) [reply crates/ruff_python_formatter/src/comments/placement.rs:1203]: I think I'd suggest treating these as dangling, and then rendering them as leading comments within the node.

Comment (charliermarsh) [reply crates/ruff_python_formatter/src/comments/placement.rs:1203]: (Leading is more consistent with how we treat comments in starred _expressions_.)

Comment (charliermarsh) [reply crates/ruff_python_formatter/src/comments/placement.rs]: (Tweaked this to merge.)

**Comment by harupy** [comment]
@charliermarsh Thanks for the update and merging!

**Comment by charliermarsh** [comment]
Thanks @harupy for another great contribution!

**Pull request merged**